### PR TITLE
Cache smart inventory memberships to avoid slow query

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
This change avoids running the host_filter query (given by the user) on every event, by running it once at the start of every job that uses a smart inventory.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
4.0.0
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: awx/main/tasks.py [hunk 1]
```diff
@@ -601,26 +601,39 @@ def update_inventory_computed_fields(inventory_id, should_update_hosts=True):
...
+                    for host_id in additions
                 ]
-                memberships.extend(add_for_inventory)
-                if add_for_inventory:
-                    changed_inventories.add(smart_inventory)
-            SmartInventoryMembership.objects.bulk_create(memberships)
-    except IntegrityError as e:
-        logger.error("Update Host Smart Inventory Memberships failed due to an exception: {}".format(e))
-        return
+                SmartInventoryMembership.objects.bulk_create(add_for_inventory)
```
Comment (matburt) [inline awx/main/tasks.py:618]: TIL about `bulk_create`

**Review by matburt — APPROVED** [review]
Looks good to me.

**Comment by AlanCoding** [comment]
Note: we will need a release note that we are going to be re-computing the smart inventory memberships more often. I don't think many people were using this, but it was technically user-facing.

**Review by ryanpetrello — APPROVED** [review]

**Comment by AlanCoding** [comment]
hey, sorry for the churn, but this particular case was bothering me:

 - start running a job against a smart inventory
 - delete some of the hosts returned by the host_filter query

In the prior form of this PR, that failed that save the event

```
awx_1        | 2019-06-11 14:56:59,453 ERROR    awx.main.commands.run_callback_receiver Callback Task Processor Raised Exception: AttributeError("'NoneType' object has no attribute 'host_id'",)
awx_1        | 2019-06-11 14:56:59,457 ERROR    awx.main.commands.run_callback_receiver Detail: Traceback (most recent call last):
awx_1        |   File "/awx_devel/awx/main/dispatch/worker/callback.py", line 107, in perform_work
awx_1        |     _save_event_data()
awx_1        |   File "/awx_devel/awx/main/dispatch/worker/callback.py", line 45, in _save_event_data
awx_1        |     cls.create_from_data(**body)
awx_1        |   File "/awx_devel/awx/main/models/events.py", line 325, in create_from_data
awx_1        |     job_event = cls.objects.create(**kwargs)
awx_1        |   File "/venv/awx/lib64/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
awx_1        |     return getattr(self.get_queryset(), name)(*args, **kwargs)
awx_1        |   File "/venv/awx/lib64/python3.6/site-packages/django/db/models/query.py", line 394, in create
awx_1        |     obj.save(force_insert=True, using=self.db)
awx_1        |   File "/awx_devel/awx/main/models/events.py", line 356, in save
awx_1        |     host_id = membership_qs.first().host_id
awx_1        | AttributeError: 'NoneType' object has no attribute 'host_id'
```

I've reproduced this situation, and then tested to confirm it works with the updated code that I just pushed. In the non-smart case, it saves the event with a null `host` field, and that it what it will do now for the smart inventory case.

**Pull request merged**